### DISCUSSION TITLE
Custom args

### DISCUSSION THREAD

**Discussion by pythoninthegrass** [raiyanyahya/freshenv#29] [Ideas]
Slick docker wrapper you've got here. Thanks for making this 🙏 

**Is your feature request related to a problem? Please describe.**
Ish. Not sure how to customize my docker image without overwriting the python source code.

**Describe the solution you'd like**
Extra click/rich flags to customize the build a la `-c customize`, `-s shell`, `-p packages`, `-o os` (or upstream image)
```bash
fr provision -f freshenv-scratch -c -o ubuntu:latest -s bash -n snek -p build-essential automake make cmake sudo git python3-pip curl vim
```

**Describe alternatives you've considered**
Could just fork the repo and do the thing. Looks simple enough, but why reinvent the wheel when you've done a good job already?

**Additional context**
A dedicated wiki could be useful if some of this is already possible and not in the readme.

**Comment by raiyanyahya** [comment]
@pythoninthegrass I am working on the exact same feature. Rather then taking in options, users can edit a freshenv.yml and build custom environments. It will look like this -

[ myenv]
Base="ubuntu18"
Install="python3-pip, wget, httppie, openjdk8, cmake ..."
Include="/home/user/startup.sh"
Run="sh startup.sh"


And you run it as 
_fr provision -f myenv_


Hope this helps :) Thank you for the interest.

**Reply by raiyanyahya** [reply]
I have also moved this to a discussion board.

**Reply by pythoninthegrass** [reply]
That looks lovely! Let me know if you need help testing ☺️

**Reply by raiyanyahya** [reply]
So, the first version with custom environments has been released :)

You can now run `fr build myenv` which will pick up the config from ~/.freshenv/freshenv.
A sample freshenv config looks like:

```console
[myenv]
base=ubuntu
install=apt update -y && apt upgrade -y && apt install snapd
cmd=bash
```

More config options will be added in the upcoming weeks. I have started working on the cloud save and share features.

Thanks.

**Reply by pythoninthegrass** [reply]
Thanks for adding that @raiyanyahya! Followed your example and got these results:
```bash
λ freshenv build config
📇 No config file found. Creating an empty config at /Users/<snip>/.freshenv/freshenv.

λ fr build myenv
🎉 Successfully built custom flavour myenv. You can provision it by running freshenv -provision -f
myenv.

λ freshenv -provision -f myenv
Usage: freshenv [OPTIONS] COMMAND [ARGS]...
Try 'freshenv --help' for help.

Error: No such option: -p

λ freshenv provision -f myenv
❌ Flavour could not be built or found.

λ fr view
💻 No freshenv environments found.

λ fr flavours
🔍 Found 5 flavours:
[
│   {
│   │   'name': 'base',
│   │   'os': 'ubuntu 18.04',
│   │   'primary-lang': 'python3',
│   │   'size': '260MB',
│   │   'shell': 'zsh',
│   │   'user': 'devuser',
│   │   'packages': ['wget git python3-pip curl zsh wget nano lsof'],
│   │   'applications': []
│   },
│   {
│   │   'name': 'py',
│   │   'os': 'debian 11',
│   │   'primary-lang': 'python3.9',
│   │   'size': '280MB',
│   │   'shell': 'zsh',
│   │   'user': 'devuser',
│   │   'packages': ['git python3-pip curl zsh wget nano lsof'],
│   │   'applications': []
│   },
│   {
│   │   'name': 'go',
│   │   'os': 'debian 11',
│   │   'primary-lang': 'golang latest',
│   │   'size': '700MB',
│   │   'shell': 'zsh',
│   │   'user': 'devuser',
│   │   'packages': ['git curl zsh wget nano lsof'],
│   │   'applications': []
│   },
│   {
│   │   'name': 'devenv',
│   │   'os': 'ubuntu latest',
│   │   'primary-lang': 'python3, node, java, golang',
│   │   'size': '1.6GB',
│   │   'shell': 'zsh',
│   │   'user': 'devuser',
│   │   'packages': [
│   │   │   'build-essential automake make cmake sudo g++ wget git python3-pip curl zsh wget nano nodejs npm fonts-powerline lsof'
│   │   ],
│   │   'applications': ['docker vscode-server']
│   },
│   {
│   │   'name': 'freshenv-scratch',
│   │   'os': 'Unix-like',
│   │   'primary-lang': [],
│   │   'size': '6mb',
│   │   'shell': 'ash',
│   │   'user': 'devuser',
│   │   'packages': [],
│   │   'applications': []
│   }
]
```

Can open an issue if you'd like?

**Reply by raiyanyahya** [reply]
Thanks :) yes you can open an issue. 
Could you tell me the version of freshenv you are running and your freshenv config.

**Reply by pythoninthegrass** [reply]
Created https://github.com/raiyanyahya/freshenv/issues/36 🙏

**Comment by pythoninthegrass** [comment]
Thanks again for your help @raiyanyahya!

Now that you [fixed](https://github.com/raiyanyahya/freshenv/issues/36) the jinja escaping error and cleaned up the success output, had another idea while testing involving whether or not the `freshenv` config could point to an existing `Dockerfile` or `docker-compose.yml`?

Was able to get around the former to some extent with the `install` line, but that would only work for simple images; to shave off a massive amount of cruft (~600 MB), I use a [builder and runner image](https://github.com/pythoninthegrass/python_template/blob/main/Dockerfile) for my Python containers.

Would be super useful to have the option to pass those files while using `freshenv` to avoid writing shell script wrappers for Docker commands 🙏

**Reply by pythoninthegrass** [reply]
[Found it](https://github.com/raiyanyahya/freshenv/blob/master/freshenv/provision.py#L51)! This is brilliant work, sir!

Already forked https://github.com/raiyanyahya/freshenv-flavours and will be playing with my custom flavors (Dockerfiles) tomorrow 🔥

**Reply by raiyanyahya** [reply]
Thank you ! Glad you liked it :)
